File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Features/src/Search/FSharpSearchFilter.fs
namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Search

open JetBrains.ProjectModel
open JetBrains.ReSharper.Plugins.FSharp
open JetBrains.ReSharper.Plugins.FSharp.Checker
open JetBrains.ReSharper.Plugins.FSharp.Psi
open JetBrains.ReSharper.Psi
open JetBrains.ReSharper.Psi.ExtensionsAPI.Finder
open JetBrains.ReSharper.Psi.Modules

type FSharpSearchFilterKey =
    { DeclaredElement: IDeclaredElement
      PsiModule: IPsiModule
      FileIndex: int }


[<PsiComponent>]
type FSharpSearchFilter(fsProjectOptionsProvider: IFcsProjectProvider) =
    let getTypeElement (fsElement: IFSharpDeclaredElement) =
        match fsElement with
        | :? ITypeElement as typeElement -> typeElement
        | fsElement -> fsElement.GetContainingType()

    interface ISearchFilter with
        member x.IsAvailable _ = true

        member x.Kind = SearchFilterKind.Cache

        member x.TryGetKey(element) =
            let fsDeclaredElement = element.As<IFSharpDeclaredElement>()
            if isNull fsDeclaredElement then null else

            let typeElement = getTypeElement fsDeclaredElement
            if isNull typeElement then null else

            let sourceFiles = typeElement.GetSourceFiles()
            if sourceFiles.IsEmpty then null else

            let declarationFileIndex =
                sourceFiles.ReadOnlyList()
                |> Seq.map fsProjectOptionsProvider.GetFileIndex
                |> Seq.min

            if declarationFileIndex = -1 then null else

            { DeclaredElement = element
              PsiModule = fsDeclaredElement.Module
              FileIndex = declarationFileIndex } :> _


        member x.CanContainReferences(sourceFile, elementId) =
            if not (sourceFile.LanguageType.Is<FSharpProjectFileType>()) then true else

            let fsElementId = elementId :?> FSharpSearchFilterKey

            let typePrivateMember = fsElementId.DeclaredElement.As<ITypePrivateMember>()
            if isNotNull typePrivateMember then
                typePrivateMember.GetSourceFiles().Contains(sourceFile) else

            if sourceFile.PsiModule != fsElementId.PsiModule then true else

            let sourceFileIndex = fsProjectOptionsProvider.GetFileIndex(sourceFile)
            sourceFileIndex >= fsElementId.FileIndex

File: ReSharper.FSharp/src/FSharp/FSharp.Psi/src/Searching/FSharpSearchFilter.cs
using System.Linq;
using JetBrains.Application.Parts;
using JetBrains.ProjectModel;
using JetBrains.ReSharper.Plugins.FSharp.Checker;
using JetBrains.ReSharper.Psi;
using JetBrains.ReSharper.Psi.ExtensionsAPI.Finder;
using JetBrains.ReSharper.Psi.Modules;
using JetBrains.ReSharper.Psi.Search;
using JetBrains.Util.DataStructures;

namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Searching
{
  [PsiComponent(Instantiation.DemandAnyThreadUnsafe)]
  public sealed class FSharpSearchFilter(IFcsProjectProvider fsProjectProvider) : ISearchFilter
  {
    private record FSharpSearchFilterKey(
      bool IsTypePrivateMember,
      IPsiModule PsiModule,
      HybridCollection<IPsiSourceFile> SourceFiles,
      int FileIndex);

    public SearchFilterKind Kind => SearchFilterKind.Cache;

    public bool IsAvailable(SearchPattern pattern) => true;

    public static bool CanContainReference(IDeclaredElement declaredElement, IPsiSourceFile sourceFile, IFcsProjectProvider fsProjectProvider)
    {
      if (!sourceFile.LanguageType.Is<FSharpProjectFileType>()) return true;

      var key = TryGetKey(declaredElement, fsProjectProvider);

      if (key.SourceFiles.Contains(sourceFile)) return true;
      if (key.IsTypePrivateMember) return false;
      if (!sourceFile.PsiModule.Equals(key.PsiModule)) return true;

      var sourceFileIndex = fsProjectProvider.GetFileIndex(sourceFile);
      return sourceFileIndex >= key.FileIndex;
    }

    private static FSharpSearchFilterKey TryGetKey(IDeclaredElement declaredElement, IFcsProjectProvider fsProjectProvider)
    {
      if (declaredElement is not IFSharpDeclaredElement fsDeclaredElement) return null;

      var typeElement = fsDeclaredElement as ITypeElement
                        ?? fsDeclaredElement.GetContainingType();

      if (typeElement is null) return null;

      var sourceFiles = typeElement.GetSourceFiles();
      if (sourceFiles.IsEmpty) return null;

      if (declaredElement is ITypePrivateMember)
        return new FSharpSearchFilterKey(IsTypePrivateMember: true, fsDeclaredElement.Module, sourceFiles, FileIndex: -1);

      var fileIndex = sourceFiles.Select(fsProjectProvider.GetFileIndex).Min();

      return new FSharpSearchFilterKey(IsTypePrivateMember: false, fsDeclaredElement.Module, sourceFiles, fileIndex);
    }

    public object TryGetKey(IDeclaredElement declaredElement) =>
      TryGetKey(declaredElement, fsProjectProvider);

    public bool CanContainReferences(IPsiSourceFile sourceFile, object key)
    {
      if (!sourceFile.LanguageType.Is<FSharpProjectFileType>()) return true;

      var fsSearchFilterKey = (FSharpSearchFilterKey)key;

      if (fsSearchFilterKey.SourceFiles.Contains(sourceFile)) return true;
      if (fsSearchFilterKey.IsTypePrivateMember) return false;
      if (!sourceFile.PsiModule.Equals(fsSearchFilterKey.PsiModule)) return true;

      var sourceFileIndex = fsProjectProvider.GetFileIndex(sourceFile);
      return sourceFileIndex >= fsSearchFilterKey.FileIndex;
    }
  }
}
